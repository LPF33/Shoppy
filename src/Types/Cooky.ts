// Types for Cooky View & Components

export interface IMeal {
  key?: string;
  meal?: string;
  startDate: string;
  endDate: string;
}

export interface IComputedMealList {
  key: string;
  meal: string;
  startDate: string;
  endDate: string;
  color?: string;
  startNum?: number;
  endNum?: number;
}

export interface INewMeal {
  startDate: Date;
  startNum: number;
}

export interface ICookyAddMealData {
  today: Date;
  startDate: null | string;
  endDate: null | string;
  meal: string;
  existingKey: false | string;
}

export interface ICookyCalendarData {
  weekday: ["Mo", "Di", "Mi", "Do", "Fr", "Sa", "So"];
  today: Date;
}
