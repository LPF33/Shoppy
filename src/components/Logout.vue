<template>
  <div id="logout-modal">
    <h2>Logout</h2>
    <h2>Wirklich ausloggen?</h2>
    <div>
      <button @click="userLogout" class="buttons">✔</button>
      <button @click="closeLogout" class="buttons">❌</button>
    </div>
  </div>
</template>

<script>
export default {
  name: "Logout",
  methods: {
    userLogout() {
      this.$store.commit("unsubscribeFirebase");
      this.$store.commit("logout");
    },
    closeLogout() {
      this.$emit("close-logout");
    },
  },
};
</script>

<style scoped>
#logout-modal {
  position: absolute;
  left: 50%;
  top: 50%;
  transform: translate(-50%, -50%);
  width: 100vw;
  padding: 20px;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;
  border-radius: 10px;
  background-color: #232323;
  color: white;
}

h1,
h2 {
  margin-bottom: 10px;
}

.buttons {
  background-color: transparent;
  font-size: 3rem;
  height: 6rem;
  width: 6rem;
  background-color: white;
  border-radius: 50%;
  line-height: 1.5;
  padding: 10px;
  cursor: pointer;
  user-select: none;
  text-decoration: none;
  margin: 20px;
}
</style>
