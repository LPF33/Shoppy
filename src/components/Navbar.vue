<template>
  <nav v-if="isLoggedIn">
    <router-link :to="{ name: 'Main' }">Main</router-link>
    <router-link :to="{ name: 'Shoppy' }">Shoppy</router-link>
    <router-link :to="{ name: 'Cook' }">Cooky</router-link>
  </nav>
  <nav v-else>
    <router-link :to="{ name: 'Login' }">Login</router-link>
  </nav>
</template>

<script>
export default {
  name: "Navbar",
  computed: {
    isLoggedIn: function () {
      return this.$store.state.isAuthenticated;
    },
  },
};
</script>

<style>
nav {
  width: 100%;
  padding: 10px;
  display: flex;
  justify-content: center;
  background: #232323;
}

nav a {
  font-weight: bold;
  color: #2c3e50;
  border: 2px solid rgb(184, 184, 184);
  padding: 10px;
  width: 150px;
  text-align: center;
  margin: 0 10px;
  color: #ffeda3;
  text-decoration: none;
}

nav a.router-link-exact-active {
  color: #39a9cb;
}
</style>
