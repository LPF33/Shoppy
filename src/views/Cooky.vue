<template>
  <main>
    <h1>👩‍🍳 👨‍🍳</h1>
    <h1>Wann kochen wir was?</h1>
    <CookyCalendar
      v-show="!modalVisible"
      :meals="mealList"
      @open-modal="openCookyAddMeal"
    />
    <CookyAddMeal
      v-show="modalVisible"
      @click.self="modalVisible = false"
      @close="modalVisible = false"
      :data-object="mealData"
    />
  </main>
</template>

<script>
import CookyCalendar from "@/components/CookyCalendar.vue";
import CookyAddMeal from "@/components/CookyAddMeal.vue";

export default {
  name: "Cooky",
  components: {
    CookyCalendar,
    CookyAddMeal,
  },
  data() {
    return {
      modalVisible: false,
      mealData: null,
    };
  },
  methods: {
    getMealDay(date) {
      date = new Date(date);
      return date.getDate();
    },
    openCookyAddMeal(meal) {
      this.mealData = meal;
      this.modalVisible = true;
    },
  },
  computed: {
    mealList() {
      return this.$store.state.mealList.map((item) => {
        const randomNum = Math.floor(Math.random() * (361 - 200) + 200);
        item.color = `hsl(${randomNum}, 80%, 70%)`;
        item.startNum = this.getMealDay(item.startDate);
        item.endNum = this.getMealDay(item.endDate);
        return item;
      });
    },
  },
};
</script>

<style scoped>
main {
  position: relative;
  background-color: #39a9cb;
  width: 100%;
  height: 100%;
  color: #232323;
  text-align: center;
}

h1 {
  margin-bottom: 10px;
}
</style>
