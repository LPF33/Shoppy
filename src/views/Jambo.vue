<template>
  <section>
    <h1>Jambo</h1>
    <h2 v-if="!counter">Start</h2>
    <h2 v-else>Spielzug: {{ counter }}</h2>
    <div v-if="visual">
      <div v-for="(item, index) in visual" :key="index" class="runde">
        <font-awesome-icon
          icon="fa-duotone fa-card-heart"
          style="
            --fa-secondary-opacity: 1;
            --fa-primary-color: rgb(255, 92, 92);
            --fa-secondary-color: white;
          "
        />
      </div>
    </div>
    <button @click="increment">
      <font-awesome-icon
        icon="fa-duotone fa-cards-blank"
        style="
          --fa-secondary-opacity: 1;
          --fa-primary-color: rgb(246, 90, 131);
          --fa-secondary-color: rgb(127, 183, 126);
        "
      />
    </button>
    <button @click="reset">
      <font-awesome-icon
        icon="fa-duotone fa-eraser"
        style="
          --fa-secondary-opacity: 1;
          --fa-primary-color: red;
          --fa-secondary-color: black;
        "
      />
    </button>
  </section>
</template>

<script lang="ts">
import { defineComponent } from "vue";

export default defineComponent({
  name: "Jambo",
  data() {
    return {
      counter: 0,
    };
  },
  methods: {
    increment() {
      if (this.counter >= 5) {
        this.reset();
        return;
      }
      this.counter++;
    },
    reset() {
      this.counter = 0;
    },
  },
  computed: {
    visual(): any[] | null {
      return this.counter > 0 ? new Array(this.counter) : null;
    },
  },
});
</script>

<style scoped>
section {
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
  text-align: center;
  background-color: rgb(255, 232, 152);
  user-select: none;
}

h1,
h2 {
  margin: 10px 0;
}

button {
  display: block;
  font-size: 3rem;
  height: 6rem;
  width: 6rem;
  border: 1px solid grey;
  border-radius: 50%;
  background-color: transparent;
  line-height: 1.5;
  padding: 10px;
  cursor: pointer;
  margin-top: 2rem;
  user-select: none;
  box-shadow: 0 0 8px black;
}

.runde {
  font-size: 3rem;
}
</style>
