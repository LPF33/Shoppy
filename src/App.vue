<template>
  <div id="view">
    <router-view />
  </div>
  <Navbar />
</template>

<script>
import Navbar from "@/components/Navbar.vue";

export default {
  components: { Navbar },
  mounted() {
    this.$store.dispatch("checkAuth");
  },
  unmounted() {
    // safe is safe ...
    this.$store.commit("unsubscribeFirebase");
  },
};
</script>

<style>
* {
  font-family: Avenir, Helvetica, Arial, sans-serif;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
}

#app {
  display: flex;
  flex-direction: column;
  width: 100vw;
  height: 100vh;
}

#view {
  flex-grow: 1;
}

.error {
  background-color: red;
  text-align: center;
  font-size: 1.5rem;
  padding: 10px;
}
</style>
