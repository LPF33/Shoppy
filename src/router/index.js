import { createRouter, createWebHistory } from "vue-router";
import Main from "../views/Main.vue";
import Shoppy from "../views/Shoppy.vue";
import Jambo from "../views/Jambo.vue";
import store from "@/store";

const routes = [
  {
    path: "/",
    name: "Main",
    component: Main,
    meta: { requiresAuth: true },
  },
  {
    path: "/shoppy",
    name: "Shoppy",
    meta: { requiresAuth: true },
    // route level code-splitting
    // this generates a separate chunk (about.[hash].js) for this route
    // which is lazy-loaded when the route is visited.
    component: Shoppy,
  },
  {
    path: "/jambo",
    name: "Jambo",
    meta: { requiresAuth: true },
    component: Jambo,
  },
  {
    path: "/budget",
    name: "Budget",
    meta: { requiresAuth: true },
    // route level code-splitting
    // this generates a separate chunk (about.[hash].js) for this route
    // which is lazy-loaded when the route is visited.
    component: () => import("../views/Budget.vue"),
  },
  {
    path: "/login",
    name: "Login",
    meta: { guest: true },
    component: () => import("../views/Login.vue"),
  },
];

const router = createRouter({
  history: createWebHistory(process.env.BASE_URL),
  routes,
});

router.beforeEach((to, from, next) => {
  if (to.matched.some((record) => record.meta.requiresAuth)) {
    if (store.state.isAuthenticated) {
      next();
      return;
    }
    next("/login");
  } else {
    next();
  }
});

router.beforeEach((to, from, next) => {
  if (to.matched.some((record) => record.meta.guest)) {
    if (store.state.isAuthenticated) {
      next("/");
      return;
    }
    next();
  } else {
    next();
  }
});

export default router;
