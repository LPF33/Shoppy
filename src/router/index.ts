import { createRouter, createWebHistory } from "vue-router";
import Main from "../views/Main.vue";
import Shoppy from "../views/Shoppy.vue";
import Jambo from "../views/Jambo.vue";
import Cooky from "../views/Cooky.vue";
import store from "@/store/index";

const routes = [
  {
    path: "/",
    name: "Main",
    component: Main,
    meta: { requiresAuth: true },
  },
  {
    path: "/shoppy",
    name: "Shoppy",
    meta: { requiresAuth: true },
    // route level code-splitting
    // this generates a separate chunk (about.[hash].js) for this route
    // which is lazy-loaded when the route is visited.
    component: Shoppy,
  },
  {
    path: "/jambo",
    name: "Jambo",
    meta: { requiresAuth: true },
    component: Jambo,
  },
  {
    path: "/budget",
    name: "Budget",
    meta: { requiresAuth: true },
    // route level code-splitting
    // this generates a separate chunk (about.[hash].js) for this route
    // which is lazy-loaded when the route is visited.
    component: () => import("../views/Budget.vue"),
  },
  {
    path: "/logout",
    name: "Logout",
    meta: { requiresAuth: true },
    // route level code-splitting
    // this generates a separate chunk (about.[hash].js) for this route
    // which is lazy-loaded when the route is visited.
    component: () => import("../views/Logout.vue"),
  },
  {
    path: "/cooky",
    name: "Cooky",
    meta: { requiresAuth: true },
    component: Cooky,
  },
  {
    path: "/login",
    name: "Login",
    meta: { guest: true },
    component: () => import("../views/Login.vue"),
  },
  {
    path: "/settings",
    name: "Settings",
    meta: { requiresAuth: true },
    component: () => import("../views/Settings.vue"),
  },
];

const router = createRouter({
  history: createWebHistory(process.env.BASE_URL),
  routes,
});

router.beforeEach((to, from, next) => {
  if (to.matched.some((record) => record.meta.requiresAuth)) {
    if (store.state.isAuthenticated) {
      next();
      return;
    }
    next("/login");
  } else if (to.matched.some((record) => record.meta.guest)) {
    if (store.state.isAuthenticated) {
      next("/");
      return;
    }
    next();
  } else {
    next("/");
  }
});

export default router;
export { routes };
