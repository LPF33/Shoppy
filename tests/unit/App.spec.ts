import { mount, shallowMount } from "@vue/test-utils";
import { createRouter, createWebHistory } from "vue-router";
import { createStore } from "vuex";
import { routes } from "@/router/index";
import App from "@/App.vue";
import Login from "@/views/Login.vue";
import Main from "@/views/Main.vue";
import Shoppy from "@/views/Shoppy.vue";
import Cooky from "@/views/Cooky.vue";
import Jambo from "@/views/Jambo.vue";
import Logout from "@/views/Logout.vue";
import Navbar from "@/components/Navbar.vue";

let actions = {
  checkIfUserIsAuth: jest.fn(),
};

function createSetup(authenticated = false) {
  actions = {
    checkIfUserIsAuth: jest.fn(),
  };

  const $store = createStore({
    state() {
      return {
        isAuthenticated: authenticated,
      };
    },
    actions,
  });

  const router = createRouter({
    history: createWebHistory(),
    routes,
  });

  router.beforeEach((to, from, next) => {
    if (to.matched.some((record) => record.meta.requiresAuth)) {
      if ($store.state.isAuthenticated) {
        next();
        return;
      }
      next("/login");
    } else {
      next();
    }
  });

  router.beforeEach((to, from, next) => {
    if (to.matched.some((record) => record.meta.guest)) {
      if ($store.state.isAuthenticated) {
        next("/");
        return;
      }
      next();
    } else {
      next();
    }
  });

  return { router, $store };
}

describe("Testing the App component", () => {
  it("App is mounted correctly when user is not logged in", async () => {
    const { router, $store } = createSetup();
    router.push("/");
    await router.isReady();

    const wrapper = mount(App, {
      global: {
        plugins: [router, $store],
      },
    });

    expect(actions.checkIfUserIsAuth).toHaveBeenCalledTimes(1);
    expect(wrapper.findComponent(Navbar).exists()).toBeTruthy();
    expect(wrapper.vm.$route.path).toBe("/login");
    expect(wrapper.findComponent(Login).exists()).toBeTruthy();
    expect(wrapper.findComponent(Main).exists()).toBeFalsy();
  });

  it("App is mounted correctly when user is logged in", async () => {
    const { router, $store } = createSetup(true);
    router.push("/");
    await router.isReady();

    const wrapper = mount(App, {
      global: {
        plugins: [router, $store],
      },
    });

    expect(actions.checkIfUserIsAuth).toHaveBeenCalledTimes(1);
    expect(wrapper.findComponent(Navbar).exists()).toBeTruthy();
    expect(wrapper.vm.$route.path).toBe("/");
    expect(wrapper.findComponent(Main).exists()).toBeTruthy();
    expect(wrapper.findComponent(Login).exists()).toBeFalsy();
  });

  it("When user not logged in, trying to go to '/' redirected to '/login'", async () => {
    const { router, $store } = createSetup(false);
    router.push("/");
    await router.isReady();

    const wrapper = shallowMount(App, {
      global: {
        plugins: [router, $store],
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.vm.$route.path).toBe("/login");
  });

  it("When user logged in, trying to go to '/login' redirected to '/'", async () => {
    const { router, $store } = createSetup(true);
    router.push("/login");
    await router.isReady();

    const wrapper = shallowMount(App, {
      global: {
        plugins: [router, $store],
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.vm.$route.path).toBe("/");
  });

  it("When user logged in, navigating to '/shoppy' will render Shoppy component", async () => {
    const { router, $store } = createSetup(true);
    router.push("/shoppy");
    await router.isReady();

    const wrapper = mount(App, {
      global: {
        plugins: [router, $store],
        stubs: {
          Shoppy: true,
        },
      },
    });

    expect(wrapper.vm.$route.path).toBe("/shoppy");
    expect(wrapper.findComponent(Shoppy).exists()).toBeTruthy();
  });

  it("When user logged in, navigating to '/cooky' will render Cooky component", async () => {
    const { router, $store } = createSetup(true);
    router.push("/cooky");
    await router.isReady();

    const wrapper = mount(App, {
      global: {
        plugins: [router, $store],
        stubs: {
          Cooky: true,
        },
      },
    });

    expect(wrapper.vm.$route.path).toBe("/cooky");
    expect(wrapper.findComponent(Cooky).exists()).toBeTruthy();
  });

  it("When user logged in, navigating to '/jambo' will render Jambo component", async () => {
    const { router, $store } = createSetup(true);
    router.push("/jambo");
    await router.isReady();

    const wrapper = mount(App, {
      global: {
        plugins: [router, $store],
        stubs: {
          Jambo: true,
        },
      },
    });

    expect(wrapper.vm.$route.path).toBe("/jambo");
    expect(wrapper.findComponent(Jambo).exists()).toBeTruthy();
  });

  it("When user logged in, navigating to '/logout' will render Logout component", async () => {
    const { router, $store } = createSetup(true);
    router.push("/logout");
    await router.isReady();

    const wrapper = mount(App, {
      global: {
        plugins: [router, $store],
        stubs: {
          Logout: true,
        },
      },
    });

    expect(wrapper.vm.$route.path).toBe("/logout");
    expect(wrapper.findComponent(Logout).exists()).toBeTruthy();
  });
});
