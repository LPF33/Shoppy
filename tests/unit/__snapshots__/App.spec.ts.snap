// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the App component When user logged in, trying to go to '/login' redirected to '/' 1`] = `
<main>
  <router-view-stub name="default"></router-view-stub>
</main>
<navbar-stub></navbar-stub>
`;

exports[`Testing the App component When user not logged in, trying to go to '/' redirected to '/login' 1`] = `
<main>
  <router-view-stub name="default"></router-view-stub>
</main>
<navbar-stub></navbar-stub>
`;
